undertow/src/main/java/org/wildfly/extension/undertow/filters/ModClusterDefinition.java

```diff
@@ -325,26 +335,72 @@ public HttpHandler createHttpHandler(Predicate predicate, ModelNode model, HttpH
 
     }
 
+    @Override
+    public void registerOperations(ManagementResourceRegistration parent) {
+        ResourceDescriptor descriptor = new ResourceDescriptor(this.getResourceDescriptionResolver())
+                .addCapabilities(Capability.class)
+                .addRequiredSingletonChildren(SingleAffinityResourceDefinition.PATH)
+                ;
+
+        for (AttributeDefinition attribute : ATTRIBUTES) {
+            descriptor.addAttributes(new Attribute() {
+                @Override
+                public AttributeDefinition getDefinition() {
+                    return attribute;
+                }
+            });
+        }
+
+        ModClusterResourceServiceHandler handler = new ModClusterResourceServiceHandler();
+        new ModClusterResourceRegistration(descriptor, handler).register(parent);
+
+    }
+
+    class ModClusterResourceRegistration extends ResourceRegistration {
+
+        ModClusterResourceRegistration(AddStepHandlerDescriptor descriptor, ResourceServiceHandler handler) {
+            super(descriptor, handler, new ModClusterAddStepHandler(descriptor,handler), new RemoveStepHandler(descriptor, handler));
+        }
+    }
+
+
     @Override
     public void registerChildren(ManagementResourceRegistration resourceRegistration) {
         resourceRegistration.registerSubModel(ModClusterBalancerDefinition.INSTANCE);
+
+        resourceRegistration.registerSubModel(new NoAffinityResourceDefinition());
+        resourceRegistration.registerSubModel(new SingleAffinityResourceDefinition());
+        resourceRegistration.registerSubModel(new RankedAffinityResourceDefinition());
     }
 
-    static class ModClusterAdd extends AbstractAddStepHandler {
-        ModClusterAdd() {
-            super(ATTRIBUTES);
+    static class ModClusterResourceServiceHandler implements ResourceServiceHandler {
+
+        @Override
+        public void installServices(OperationContext context, ModelNode model) throws OperationFailedException {
+            String name = context.getCurrentAddressValue();
+            ModelNode recursiveModel = Resource.Tools.readModel(context.readResourceFromRoot(context.getCurrentAddress(), true));
+            ModClusterService.install(name, context.getCapabilityServiceTarget(), recursiveModel, context);
         }
 
         @Override
-        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
-            final String name = context.getCurrentAddressValue();
-            ModClusterService.install(name, context.getCapabilityServiceTarget(), model, context);
+        public void removeServices(OperationContext context, ModelNode model) {
+            context.removeService(UndertowService.FILTER.append(Constants.MOD_CLUSTER));
+        }
+    }
+
+    static class ModClusterAddStepHandler extends AddStepHandler {
```

Comment:
There are better ways to provide a custom Resource implementation without needing to extend AddStepHandler, but they will require refactoring ModClusterResource itself.  But we can address this separately.